presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -98,6 +97,9 @@ public PlanNodeDecorrelator(SymbolAllocator symbolAllocator, Lookup lookup)
         @Override
         protected Optional<DecorrelationResult> visitPlan(PlanNode node, Void context)
         {
+            if (containsCorrelation(node, correlation)) {
```

Comment:
I put the check here to enable early failure -- and collapse the recursion quickly with `Optional.empty()` instead of passing a `DecorrelationResult` and triggering decorrelation routines for all the nodes above just to fail at the final check (in `decorrelatedNode()` method).

Reply:
I would remove this check since its redundant, but brings confusion to the code. The performance gain is insignificant.

Alternatively we could make the other check to be `checkState`.